File: python_workflow/tests/helper_repo.py
"""Locating the lexsync repository around the package, for the tests that hold the
package to files kept only in the repository: templates/, config/, items/, corpora/,
output/, apps/, the R package and the citation metadata at the root.

From a checkout the tests run in python_workflow/tests, two levels below the root.
The sdist ships tests/ as well, for conda-forge, Debian and Nix to build from (see
pyproject.toml), and there the same walk ends in whatever directory the sdist was
unpacked into, which can hold anything. A bare name can therefore be somebody
else's file: the R twin's test for templates/ failed CRAN's Linux checks on
finding the unrelated CRAN package of that name. A candidate root counts only if
it carries the R package's own DESCRIPTION under R_workflow/ and python_workflow/
beside it, which is the check R_workflow/tests/testthat/helper-repo.R makes.
Anywhere else REPO is None, and every test that needs it skips.
"""
import re
from pathlib import Path

import pytest


def repo_root():
    """The repository root, or None when the package is not inside its repository."""
    # In the repository this file sits at python_workflow/tests/, so it always has
    # three parents. A shallower path, such as tests/ unpacked straight into / or
    # C:\, cannot be the repository, and indexing past the root there would raise
    # at import and halt collection of the whole suite.
    parents = Path(__file__).resolve().parents
    if len(parents) < 3:
        return None
    root = parents[2]
    desc = root / "R_workflow" / "DESCRIPTION"
    if not desc.is_file() or not (root / "python_workflow").is_dir():
        return None
    try:
        text = desc.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return None
    package = re.search(r"^Package:[ \t]*(\S+)", text, re.MULTILINE)
    return root if package and package.group(1) == "lexsync" else None


# Resolved once, at import, because some modules build their paths and parameters
# from it at collection time.
REPO = repo_root()


def repo_path(*parts, reason="the lexsync repository is not around this package",
              required=False):
    """A path under the repository root, skipping the calling test with `reason`
    when the package is not inside its repository. With `required` the test fails
    instead, for a job that set out to run against the repository."""
    # Report the skip or failure at the caller, which names the test, not here.
    __tracebackhide__ = True
    if REPO is None:
        if required:
            pytest.fail(reason)
        pytest.skip(reason)
    return REPO.joinpath(*parts)

File: R_workflow/tests/testthat/helper-repo.R
# Locating the lexsync repository around the package, for the tests that hold the
# package to files kept only at the repository root: templates/, config/, items/,
# apps/ and README.md.
#
# From a source checkout the tests run in R_workflow/tests/testthat, three levels
# below the root; a local R CMD check started in R_workflow/ runs them from
# R_workflow/lexsync.Rcheck/tests/testthat, four levels below it, and reaches the
# repository only with NOT_CRAN=true, as devtools::check() sets. On CRAN the same
# walk ends wherever the check was started, and on its Linux hosts that directory
# holds the unpacked sources of other CRAN packages. A bare name can therefore be
# somebody else's package: the 'templates' package once passed a dir.exists() guard
# here and failed the check. A candidate root counts only if it carries this
# package's own DESCRIPTION under R_workflow/ and the Python twin beside it, and on
# CRAN every test that needs it skips, since no repository can be present there.

# The repository root, or NULL when the package is not inside its repository.
repo_root <- function() {
  for (depth in 3:4) {
    root <- normalizePath(do.call(testthat::test_path, as.list(rep("..", depth))),
                          winslash = "/", mustWork = FALSE)
    desc <- file.path(root, "R_workflow", "DESCRIPTION")
    if (!file.exists(desc) || !dir.exists(file.path(root, "python_workflow"))) next
    pkg <- tryCatch(read.dcf(desc, fields = "Package")[1, 1], error = function(e) NA)
    if (identical(unname(pkg), "lexsync")) return(root)
  }
  NULL
}

# A path under the repository root, skipping the calling test with `message` on
# CRAN or when the package is not inside its repository.
repo_path <- function(..., message = "the lexsync repository is not around this package") {
  testthat::skip_on_cran()
  root <- repo_root()
  if (is.null(root)) testthat::skip(message)
  file.path(root, ...)
}
